### PR TITLE
frogger to review

### PR DESCRIPTION
**Pull request by kalash14**
_Complete the form below and remove this line._

### Object-Oriented JavaScript Frogger Game

[Demo](https://kalash14.github.io/frogger-game/) |
[Code base](https://github.com/kalash14/frogger-game)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/kalash14/js-oop-frogger/js/app.js [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    }
+    if(this.x < 0){
+        this.x = (boundaryRight - blockWidth);
+    }
+
+};
+
+/* detecting collision between enemies and player */
+
+let checkCollision = function(enemy, player) {
```
Comment (OleksiyRudenko) [inline submissions/kalash14/js-oop-frogger/js/app.js]: Anyway you've hacked game engine. Consider making this and other global functions a part of engine.

You may also opt for making collision check a part of Enemy behaviour. Just let each enemy know about player at each enemy instantiation.

You are not require to fix the above.

File: submissions/kalash14/js-oop-frogger/js/app.js [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+};
+
+/* defining enemies and player */
+
+let firstEnemy = new Enemy(-blockWidth, 60),
+    secondEnemy = new Enemy(-blockWidth, 145),
+    thirdEnemy = new Enemy(-blockWidth, 220),
+    allEnemies = [firstEnemy, secondEnemy, thirdEnemy],
+    player = new Player(startX,startY);
+
```
Comment (OleksiyRudenko) [inline submissions/kalash14/js-oop-frogger/js/app.js]: How about `allEnemies = [1,2,3].map(fieldRow => new Enemy(-blockWidth, fieldRow * blockHeight - fieldTopOffsetPx))` (where `fieldTopOffsetPx === 20` or so)?

You will get rid of a couple of extra magic numbers and also easily scale initial number of enemies.

File: submissions/kalash14/js-oop-frogger/js/app.js [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    if (score > highScore) {
+        highScore = score;
+        highScoreElement.innerHTML = highScore;
+    }
+
+};
+
+Player.prototype.render = function() {
+    let imageElement = document.createElement('img');
```
Comment (OleksiyRudenko) [inline submissions/kalash14/js-oop-frogger/js/app.js]: Do you really need creating an `img` element at each render? This is a quite expensive operation, and also enemies get rendered in a somehow simpler method.

Basically, it seems character rendering should not differ among characters.

Therefore you might unify this method and then generalize this into a base class.

File: submissions/kalash14/js-oop-frogger/js/app.js [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+/* changing disable state of START button */
+
+let checkStartButton = () => {
+
+    if (notEmptyInputField && playerImageSrc !== "") {
+        startGameButton.disabled = false;
+    }
+    else {
+        startGameButton.disabled = true;
+    }
```
Comment (OleksiyRudenko) [inline submissions/kalash14/js-oop-frogger/js/app.js]: Consider this equivalent one-liner
`startGameButton.disabled = !(notEmptyInputField && playerImageSrc !== "")`

Your condition evaluates into boolean and the boolean is what you need to assign then. Why not combine?

Further condition optimization (inversion) is also recommended as normally `!(condition)` is counter-intuitive in such cases.

File: submissions/kalash14/js-oop-frogger/js/app.js [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    userNameBlock = document.querySelector('.game-session__user'),
+    startModal = document.querySelector('.start-game-modal'),
+    copyRightSection = document.querySelector('.copyright-section'),
+    audioTrack = document.querySelector('.audio-track'),
+    soundSwitcher = document.querySelector('.sound-switcher'),
+    gameSession = document.querySelector('.game-session'),
+    playersGrid = document.querySelector('.players-row'),
+    playerScoreElement = document.querySelector('.game-session__score'),
+    highScoreElement = document.querySelector('.game-session__maxscore');
+
```
Comment (OleksiyRudenko) [inline submissions/kalash14/js-oop-frogger/js/app.js]: Consider aggregating the above constants like following:
```
const UI_ELEMENTS = {
  userNameInput: document.querySelector('.start-game-input'),
  startGameButton: document.querySelector('.start-game-button'),
  // etc.
};
```

You are not required to fix this.

**Comment by OleksiyRudenko** [comment]
@kalash14 Please, resolve conversation for the comments you've fixed. This helps to understand if you consider anything yet to add/change before next review iteration.
For the parts you're not going to fix, please, give a reason.
Thanks.

File: submissions/kalash14/js-oop-frogger/js/app.js (hunk 5 above)
Comment (kalash14) [reply submissions/kalash14/js-oop-frogger/js/app.js]: Thanks! Fixed!

File: submissions/kalash14/js-oop-frogger/js/app.js (hunk 4 above)
Comment (kalash14) [reply submissions/kalash14/js-oop-frogger/js/app.js]: Thanks! Fixed!

File: submissions/kalash14/js-oop-frogger/js/app.js (hunk 3 above)
Comment (kalash14) [reply submissions/kalash14/js-oop-frogger/js/app.js]: Thanks! Fixed!

File: submissions/kalash14/js-oop-frogger/js/app.js (hunk 1 above)
Comment (kalash14) [reply submissions/kalash14/js-oop-frogger/js/app.js]: Thanks! Fixed!

**Comment by kalash14** [comment]
@OleksiyRudenko, thanks for reviewing! I have fixed some parts of my code according to Your recommendations and also resolved comments above.

File: submissions/kalash14/js-oop-frogger/js/app.js [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    userNameBlock: document.querySelector('.game-session__user'),
+    startModal: document.querySelector('.start-game-modal'),
+    copyRightSection: document.querySelector('.copyright-section'),
+    audioTrack: document.querySelector('.audio-track'),
+    soundSwitcher: document.querySelector('.sound-switcher'),
+    gameSession: document.querySelector('.game-session'),
+    playersGrid: document.querySelector('.players-row'),
+    playerScoreElement: document.querySelector('.game-session__score'),
+    highScoreElement: document.querySelector('.game-session__maxscore')
+};
```
Comment (OleksiyRudenko) [inline submissions/kalash14/js-oop-frogger/js/app.js]: You are not required to fix this any further. Just a yet another possible approach for your consideration.

This part could be optimized event further, e.g.
```
const UI_ELEMENTS = {
  userNameInput: '.start-game-input',
  startGameButton: '.start-game-button',
  // etc.
};
for (let key in UI_ELEMENTS) {
  UI_ELEMENTS[key] = document.querySelector(UI_ELEMENTS[key]);
}
```
The above can be optimized even further and the operation is probably worth generalizing into a specialized service function if applicable 1+ times across code.

e.g.  (a bit verbose function name just to give a clear idea of its purpose)
```
const UI_ELEMENTS = applyTransformerToObjectValues({
  //...
}, document.querySelector);

function applyTransformerToObjectValues(objectToTraverse, transformFunction) {
  for (let key in objectToTraverse) {
    objectToTraverse[key] = transformFunction(objectToTraverse[key]);
  }
}
```

Comment (kalash14) [reply submissions/kalash14/js-oop-frogger/js/app.js]: Ok!

**Pull request merged**